lib/util/traffic-switch/traffic-switcher.ts

```diff
@@ -0,0 +1,139 @@
+import { log, metric, MetricLoggerUnit } from '@uniswap/smart-order-router'
+
+/*
+  This class is a base class that can be used for traffic switching or sampling between two implementations of an interface.
+  It provides the ability to switch traffic between two implementations based on a percentage,
+  or to sample traffic between two implementations based on a percentage (one or the other, not both at the same time).
+  ExperimentName, aliasA and aliasB are unique identifiers and will be used for logging purposes.
+ */
+export abstract class TrafficSwitcher<T> {
+  public static readonly METRIC_NAME_TEMPLATE: string = 'TRAFFIC_SWITCHER__{EXP}__{METHOD}__{METRIC}'
+
+  // Used for logging purposes.
+  protected readonly experimentName: string
+  protected readonly aliasA: string
+  protected readonly aliasB: string
+  // 2 implementations of the interface that we want to switch traffic between.
+  protected readonly implementationA: T
+  protected readonly implementationB: T
+
+  // pctTrafficSwitchToB is the percentage of traffic that will be switched to implementationB
+  protected readonly pctTrafficSwitchToB: number
+  // pctSamplingToCompare is the percentage of traffic that will be compared between implementationA and implementationB
+  protected readonly pctSamplingToCompare: number
+
+  constructor(
+    experimentName: string,
+    implementationA: T,
+    implementationB: T,
+    aliasA: string,
+    aliasB: string,
+    pctTrafficSwitchToB: number = 0.0,
+    pctSamplingToCompare: number = 0.0
+  ) {
```

Comment:
Here's a very neat Typescript style that I've seen in many repos and which I think you might like, It's a pattern that the AWS CDK and SDK use, as well as SST, and I think once you get into the groove of it, it's super helpful.

```
export interface TrafficSwitcherProps<TExperiment> {
  control: TExperiment;
  treatment: TExperiment;
  aliasControl: string,
  aliasTreatment: string,
  trafficSwitch: TrafficSwitchProps;
}

export interface TrafficSwitchProps {
  pctEnabled: number;
  pctShadowSampling: number;
}
```

then in this constructor:
```
  constructor(experimentName: string, props: TrafficSwitcherProps) {
    // here you map your props to your local variables.
   }
```